lib/internal/modules/esm/hooks.js

```diff
@@ -0,0 +1,653 @@
+          'a string',
+          hookErrIdentifier,
+          preload,
+        );
+      }
+      const { compileFunction } = require('vm');
+      const preloadInit = compileFunction(
+        preloaded,
+        ['getBuiltin', 'port', 'setImportMetaCallback'],
+        {
+          filename: '<preload>',
+        }
+      );
+      const { BuiltinModule } = require('internal/bootstrap/loaders');
+      // We only allow replacing the importMetaInitializer during preload;
+      // after preload is finished, we disable the ability to replace it.
+      //
+      // This exposes accidentally setting the initializer too late by throwing an error.
+      let finished = false;
+      let replacedImportMetaInitializer = false;
+      let next = this.#importMetaInitializer;
+      try {
+        // Calls the compiled preload source text gotten from the hook
+        // Since the parameters are named we use positional parameters
+        // see compileFunction above to cross reference the names
+        FunctionPrototypeCall(
+          preloadInit,
+          globalThis,
+          // Param getBuiltin
+          (builtinName) => {
+            if (BuiltinModule.canBeRequiredByUsers(builtinName) &&
+                BuiltinModule.canBeRequiredWithoutScheme(builtinName)) {
+              return require(builtinName);
+            }
+            throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+          },
+          // Param port
+          insidePreload,
+          // Param setImportMetaCallback
+          (fn) => {
+            if (finished || typeof fn !== 'function') {
+              throw new ERR_INVALID_ARG_TYPE('fn', fn);
+            }
+            replacedImportMetaInitializer = true;
+            const parent = next;
+            next = (meta, context) => {
+              return fn(meta, context, parent);
+            };
+          });
+      } finally {
+        finished = true;
+        if (replacedImportMetaInitializer) {
+          this.#importMetaInitializer = next;
+        }
+      }
+    }
+  }
+
+  importMetaInitialize(meta, context) {
+    this.#importMetaInitializer(meta, context);
+  }
+
+  /**
+   * Resolve the location of the module.
+   *
+   * Internally, this behaves like a backwards iterator, wherein the stack of
+   * hooks starts at the top and each call to `nextResolve()` moves down 1 step
+   * until it reaches the bottom or short-circuits.
+   *
+   * @param {string} originalSpecifier The specified URL path of the module to
+   *                                   be resolved.
+   * @param {string} [parentURL] The URL path of the module's parent.
+   * @param {ImportAssertions} [importAssertions] Assertions from the import
+   *                                              statement or expression.
+   * @returns {Promise<{ format: string, url: URL['href'] }>}
+   */
+  async resolve(
+    originalSpecifier,
+    parentURL,
+    importAssertions = ObjectCreate(null),
```

Comment:
```suggestion
    importAssertions = { __proto__: null },
```
nit, just to avoid an unnecessary function call